store/aerospike/aerospike.go

```diff
@@ -0,0 +1,322 @@
+	if err != nil {
+		return nil, err
+	}
+	err = <-regTask.OnComplete()
+	if err != nil {
+		return nil, err
+	}
+
+	s.client = client
+
+	return s, nil
+}
+
+// newKey initializes a key from namespace
+func (s *Aerospike) newKey(key string) (*api.Key, error) {
+	var akey *api.Key
+	ikey, err := strconv.ParseInt(key, 10, 64)
+	if err == nil {
+		akey, err = api.NewKey(s.namespace, "", ikey)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		akey, err = api.NewKey(s.namespace, "", key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return akey, err
+}
+
+// Get the value at "key", returns the last modified index
+// to use in conjunction to CAS calls
+func (s *Aerospike) Get(key string, opts *store.ReadOptions) (*store.KVPair, error) {
+	policy := api.NewPolicy()
+	akey, err := s.newKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	record, err := s.client.Get(policy, akey)
+	if record == nil {
+		return nil, store.ErrKeyNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &store.KVPair{Key: key, Value: record.Bins["bin"].([]byte), LastIndex: uint64(record.Generation)}, nil
+}
+
+// Put a value at "key"
+func (s *Aerospike) Put(key string, value []byte, opts *store.WriteOptions) error {
+	akey, err := s.newKey(key)
+	if err != nil {
+		return err
+	}
+
+	policy := api.NewWritePolicy(0, 0)
+	policy.SendKey = true
+	if opts != nil && opts.TTL > 0 {
+		policy.Expiration = uint32(opts.TTL / time.Second)
+	}
+	bin := api.NewBin("bin", value)
+	err = s.client.PutBins(policy, akey, bin)
+	return err
+}
+
+// Delete a value at "key"
+func (s *Aerospike) Delete(key string) error {
+	akey, err := s.newKey(key)
+	if err != nil {
+		return err
+	}
+	exists, err := s.client.Delete(nil, akey)
+	if !exists {
+		return store.ErrKeyNotFound
+	}
+	return err
+}
+
+// Exists checks that the key exists inside the store
+func (s *Aerospike) Exists(key string, opts *store.ReadOptions) (bool, error) {
+	policy := api.NewPolicy()
+
+	akey, err := s.newKey(key)
+	if err != nil {
+		return false, err
+	}
+
+	exists, err := s.client.Exists(policy, akey)
+	if !exists {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// List returns the range of keys starting with the passed in prefix
+// Aerospike does not store primary keys by default and does not support key searches...
+// Used full scan of the keys - very slowly!
+// It is not recommended to use without special need.
+func (s *Aerospike) List(keyPrefix string, opts *store.ReadOptions) ([]*store.KVPair, error) {
+	stmt := api.NewStatement(s.namespace, "")
+	stmt.SetAggregateFunction("valkeyrie", "list", []api.Value{api.NewValue(keyPrefix)}, true)
+	rs, err := s.client.Query(nil, stmt)
+	if err != nil {
+		return nil, err
+	}
+
+	kv := []*store.KVPair{}
+
+	keyFound := false
+	for res := range rs.Results() {
+		if res.Err != nil {
+			return nil, res.Err
+		}
+		record := res.Record.Bins["SUCCESS"].(map[interface{}]interface{})
```

Comment:
Same thing here I think as an added safety we shall check that a value exists at that key before trying to cast it. (Otherwise we have a nil map access below)